src/oc_erchef/apps/migrator/src/migrator_change_listener.erl

```diff
@@ -0,0 +1,211 @@
+%% -*- erlang-indent-level: 4;indent-tabs-mode: nil -*-
+%% ex: ts=4 sw=4 et
+%% @author Marc Paradise <marc@chef.io>
+%% Copyright 2017 Chef Software, Inc. All Rights Reserved.
+%%
+%% This file is provided to you under the Apache License,
+%% Version 2.0 (the "License"); you may not use this file
+%% except in compliance with the License.  You may obtain
+%% a copy of the License at
+%%
+%%   http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+
+-module(migrator_change_listener).
+-behaviour(gen_server).
+
+-define(SERVER, ?MODULE).
+
+-export([%% API functions
+         start_link/0,
+
+         %% gen_server behaviour
+         init/1,
+         handle_call/3,
+         handle_cast/2,
+         terminate/2,
+         handle_info/2,
+         code_change/3]).
+
+-define(POLL_INTERVAL_MS, 250).
+
+% This module simply polls for new transactions using a direct sql connection as superuser.
+% The end of this first phase will have it running on localhost, dumping data into a local copy of the DB.
+% Once parsing/statements are worked out, the next step is a remote DB associated with a chef server,
+% perhaps with a manual restore of bifrost and cookbooks.
+%
+% Because this is PoC/spike work, you won't find new tests here yet.
+%
+% Some thoughts:
+%  - currently this does all parsing in the same proc. We may want to farm that out to
+%    one or more workers, which can sync and sequence on the trans id.  Particularly once we're
+%    doing this for multiple DBs.
+%  - This does attempt to prevent transaction loss by 'peek'ing one change at a time, then
+%    'get'ing it to clear it from the replication slot.  Errors that cause the supervisor to give up
+%    due to restart intensity limits would be considered fatal to the migration.
+%  - I used epgsql directly:
+%       * sqerl only supports one DB connection - presently that's opscode_chef as the erchef user;
+%         as an alterntiave I could have used sqerl for the polling by giving the erchef user
+%         replication privileges, though initially the direction was a bit different and that made less sense.
+%       * we don't need pooling.  There's a dedicated connection to siphone dat from the logical repl slot
+%         and that's it.
+%       * When/if we move this to its own standalone service, that's a thing to revisit and check perf impact of.
+%
+% Open questions:
+%  - how do we handle updatesa that are done as part of triggers or stored procedures?
```

Comment:
I think the way to do this is to set the session_replication_role on the session to replica: https://www.postgresql.org/docs/9.6/static/runtime-config-client.html